server-spi-private/src/main/java/org/keycloak/device/DeviceActivityManager.java

```diff
@@ -27,22 +27,18 @@
 import org.keycloak.models.UserSessionModel;
 import org.keycloak.representations.account.DeviceRepresentation;
 import org.keycloak.util.JsonSerialization;
-import ua_parser.Client;
-import ua_parser.Parser;
+import nl.basjes.parse.useragent.UserAgentAnalyzer;
+import nl.basjes.parse.useragent.UserAgent;
 
 /**
  * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
  */
 public class DeviceActivityManager {
```

Comment:
I'd argue this should really be an SPI and not a util class like this. That would allow the factory to create a singleton provider that is re-used rather than rely on static/final in a helper class.

Reply:
+1, that should also help with the initialization being done at startup (as it can be triggered in factory `init` or `postInit`).

Still will be also good to trigger initialization in "async" way (in separate thread) IMO as I've mentioned above as increase server startup for 2-3 seconds just for the initialization of this stuff is quite a lot. Also it is not ideal to block 1st HTTP request to Keycloak for 2-3 seconds, so suggesting the initialization to be triggered at server startup rather than 1st request.

Reply:
2-3 is a very long initialization. Why is it that slow? Really seems like that is adding a lot of overhead to getting ready just to be able to parse some headers. Regardless of how it's initalized (startup, in the background, or for first request) that is a lot of wasted CPU cycles just to initialize something small like this.

Reply:
The 2-3 seconds come from turning a few thousand rules into data structures that have been optimised for execution performance. Note that if you ask the system at startup for less fields both the startup time and runtime improve.

Reply:
@stianst @mposolda did I miss something in the SPI config that would cause the current failures?